python/tests/test_pyvw.py

```diff
@@ -344,3 +346,49 @@ def check_error_raises(type, argument):
     """
     with pytest.raises(type) as error:
         argument()
+
+
+def test_oneline_simple_conversion():
+    df = pd.DataFrame({"y": [1], "x": [2]})
+    conv = DataFrameToVW(df, "{y} | {x}")
+    lines_list = conv.process_df()
+    first_line = lines_list[0]
+    assert first_line == "1 | 2"
+
+
+def test_oneline_with_column_renaming_and_tag():
+    df = pd.DataFrame({"idx": ["id_1"], "y": [1], "x": [2]})
+    conv = DataFrameToVW(df, "{y} {idx}| col_x:{x}")
+    lines_list = conv.process_df()
+    first_line = lines_list[0]
+    assert first_line == "1 id_1| col_x:2"
+
+
+def test_multiple_lines_conversion():
+    df = pd.DataFrame({"y": [1, -1], "x": [1, 2]})
+    conv = DataFrameToVW(df, "{y} | {x}")
+    lines_list = conv.process_df()
+    assert lines_list == ["1 | 1", "-1 | 2"]
+
+
+def test_oneline_with_multiple_namespaces():
+    df = pd.DataFrame({"y": [1], "a": [2], "b": [3]})
+    conv = DataFrameToVW(df, "{y} |FirstNameSpace {a} |DoubleIt:2 {b}")
```

Comment:
We could also consider an api for DataFrameToVW that takes in some sort of spec of what we want to output. Just a sketch not actual code:

```
[
 Namespace(name="FirstNamespace", features=[Feature(name=at("a"))]),
, Namespace(name="DoubleIt", value=2, features=[Feature(name=at("b"))])
]
```

Reply:
Yes @lalo, I think this is good too !
To sum up, for the following formula (which is a bit more complex to take more cases)
`"{y} {tag_col}|FirstNameSpace {a} {b}:2 |DoubleIt:2 ColC:{c} |Constant :1"`

We could have something like :
```python
DataFrameToVW(
    df, 
    Targets = [Label(col("y"), Tag(col("a"))], 
    Features = [
    Namespace(name="FirstNamespace", 
              features=[Feature(value=col("a")), Feature(name=col("b"), value=2]),
    Namespace(name="DoubleIt", value=2, 
              features=[Feature(name="colC", value=col("c"))]),
    Namespace(name="Constant"), 
              features=[Feature(value=1)]
])
```
In my opinion, the approach of using extensive args is more explicit and does not require complex regex checking. However it's more tedious to write, which can be a drawback if you have a lot of features.

Regarding the approach based on string formula, the main benefit is that is fast to write and familiar to people from statistical background that are used to R-formula like (as in R or the python package [statsmodels](https://www.statsmodels.org/stable/index.html#index--page-root)).

What do you think @jackgerrits, @lalo ?

Reply:
We could also add convenience methods for example have some sort of default mapping that assumes no namespaces or values, anything more complicated than that someone has to define the explicit mapping. That and a filtered one where you send only the columns from that frame that convert into features.

Reply:
Yeah I agree convenience methods for common ways to map it can be useful.

I find the `Targets` value a little confusing as in the text format everything to the left of the first `|` is the label (minus the tag). Therefore I think there should be a top level property which takes a Label object of a specific label type. We would need to specify all of those, but we can easily just start with `SimpleLabel` type.

You could also have the string formula generate this programmatic formula. Could be another nice extension convenience maybe? Maintaining the regex maybe not desirable though.

Reply:
Again with pseudocode to sketch out the api: 

we could have something like 
```
DataFrameToVW(
    df,
    VWMappingDefault(df.columns()))
```
or
```
DataFrameToVW(
    df,
    VWMappingDefault(filter(df.columns,["a","b"])) # in case df has more columns we don't care about
```

Then we would have to make sure that VWMappingDefault just generates the structure of having no namespace, no values, etc just grabbing whatever value from the column into a feature.

Also a benefit on leaning on extra types is that if we change or update the input format we don't break all the users of the api (i.e. changing the '|' to '#', not that this will ever happen). We would be able to transform their types into the new input format cleanly.

Reply:
>We could also add convenience methods for example have some sort of default mapping that assumes no namespaces **or values**, anything more complicated than that someone has to define the explicit mapping. 

You mean no feature name ? OK for the default mapping !

> I find the Targets value a little confusing as in the text format everything to the left of the first | is the label (minus the tag). Therefore I think there should be a top level property which takes a Label object of a specific label type. We would need to specify all of those, but we can easily just start with SimpleLabel type.

In the vowpal input format wiki `[Label] [Importance] [Base] [Tag]|Namespace Features |Namespace Features ... |Namespace Features`. So at the left of the first `|` there can also be _Importance_ and _Base_. Also, as you mentionned earlier there is also the polylabel union (for CB). 

We could defined, for the LHS of the first "|" the following properties/types : `Label`, `Importance`, `Base`, `Tag` and `UnionLabel/PolyLabel` (for the CB case) and the following properties for the RHS of the first "|": `Namespace`, `Feature`

> Again with pseudocode to sketch out the api:
> 
>  we could have something like
> 
> DataFrameToVW(
>     df,
>     VWMappingDefault(df.columns()))
> or
> 
> DataFrameToVW(
>     df,
>     VWMappingDefault(filter(df.columns,["a","b"])) # in case df has more columns we don't care about
> Then we would have to make sure that VWMappingDefault just generates the structure of having no namespace, no values, etc just grabbing whatever value from the column into a feature.

Sounds good !  The by default the first column of the DataFrame would be the target ? Alternatively we could ask to supply the column name of the target and the list of names of the features as in `VWMappingDefault(y="y", x = ["a", "b"])`

> Also a benefit on leaning on extra types is that if we change or update the input format we don't break all the users of the api (i.e. changing the '|' to '#', not that this will ever happen). We would be able to transform their types into the new input format cleanly.

Yes, I totally agree !

Reply:
`[Label] [Importance] [Base] [Tag]|` is only true for `simple_label`. Unfortunately, this is a case of the wiki not quite being up to date.

LHS of the first `|` will depend on the label type, and therefore it must be specific per label type. This is why I suggest providing a rich label object and the tag. But RHS of the first `|` will be the same for _every_ label type.

Reply:
> [Label] [Importance] [Base] [Tag]| is only true for simple_label. Unfortunately, this is a case of the wiki not quite being up to date.

Another case is the pattern for CB (`action:cost:probability | features`). Is there other cases ?

> LHS of the first | will depend on the label type, and therefore it must be specific per label type. This is why I suggest providing a rich label object and the tag. But RHS of the first | will be the same for every label type.

I have been thinking about the design and I could do the following an abstract class `FeatureHandler` with abstract method `process` and concrete method `get_col_or_value`. The following 2 classes will inherit of this abstract class : 

- `SimpleLabel`, which has an attribute **name** and implements process.
- `Feature`, which has an attribute **name** and **value** and implements process.

In both class, the attributes **name** (and **value**) can either receive an object of type **col** (that specified the column to extract as in col("a")) or a value that will be considered as it. The concrete method `get_col_or_value` will extract the column from the dataframe (if **col**) or build the column with the repeat value if a value is passed. 

The implementation of the `process` method will build the appropriate column of string (pandas.Series) according to the type of the object (SimpleLabel, Feature or PolyLabel)

A third class `PolyLabel` with attributes **action**, **cost**, **probability** can easily be added.

What do you think of it ?

Reply:
I thought of it and simplified the classes (no abstract class, and no subtypes Label/Importance/Base/Importance).

Here is my UML design of the classes I wrote :

![image](https://user-images.githubusercontent.com/17943873/81400971-96a2db80-912e-11ea-99f9-578f74d72bfe.png)

**Usages**

Let's build a toy dataset

```python
import numpy as np
import pandas as pd

df_toy = pd.DataFrame(
    {
        "y": [-1, -1, 1],
        "p": np.random.uniform(size=3),
        "c": [4.5, 6.7, 9.6],
        "a": [1, 2, 3],
        "b": np.random.normal(size=3)
    }
)
# out
   y         p    c  a         b
0 -1  0.525440  4.5  1  0.616254
1 -1  0.262586  6.7  2 -1.133934
2  1  0.705830  9.6  3 -0.452018
```

**1. Using just colnames to specify target and features**
Same idea as @lalo 's VWMappingDefault. We call the method `DFtoVW.from_colnames` :
```python 
DFtoVW.from_colnames(y="y", X =["a"], df=df_toy).process_df()
# out
['-1 | 1', '-1 | 2', '1 | 3']


DFtoVW.from_colnames(y="y", X = set(df_toy) - set(["y", "p"]), df=df_toy).process_df()
# out
['-1 | 0.6162539137336357 1 4.5',
 '-1 | -1.1339344282053312 2 6.7',
 '1 | -0.45201750087182024 3 9.6']


DFtoVW.from_colnames(y=["a", 'c', "p"], X =["a"], df=df_toy, poly_label=True).process_df()
# out
['1:4.5:0.5254401040586195 | 1',
 '2:6.7:0.2625862316861749 | 2',
 '3:9.6:0.7058299954250811 | 3']
```

**2. Using the interface that we talked about**
```python
DFtoVW(label=SimpleLabel(Col("y")), 
       namespaces=Namespace(features=[Feature(value=Col("a"))]),
       df=df_toy).process_df()
# out
['-1 | 1', '-1 | 2', '1 | 3']


complex_namespaces = [
    Namespace(name="FirstNamespace", features=[Feature(name="ColA", value=Col("a"))]),
    Namespace(name="DoubleIt", value=2, features=[Feature(value=Col("b"))])
]
DFtoVW(label=SimpleLabel(Col("y")), 
       namespaces=complex_namespaces, 
       df=df_toy).process_df()
# out
['-1 |FirstNamespace ColA:1 |DoubleIt:2 0.6162539137336357',
 '-1 |FirstNamespace ColA:2 |DoubleIt:2 -1.1339344282053312',
 '1 |FirstNamespace ColA:3 |DoubleIt:2 -0.45201750087182024']


DFtoVW(label=PolyLabel(action=Col("a"), cost=Col("c"), proba=Col("p")),
       namespaces=Namespace(features=[Feature(value=Col("b"))]),
       df=df_toy).process_df()
# out
['1:4.5:0.6121907426717043 | -0.43104977544310435',
 '2:6.7:0.6773696976137632 | 2.45363493233382',
 '3:9.6:0.5955350558877885 | 1.2190748658325201']
```

Is that class design okay for you ?